src/vm.rs

```diff
@@ -690,14 +690,63 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
                     event_data,
                 );
                 if let Ok(counter_index) = result && counter_index < pmu_info.get_num_counters() {
-                    // TODO: Update internal state for configured counters.
+                    update_configured_counter(counter_index, config_flags, event_type, active_cpu);
                 }
                 result
             } else {
                 Err(Error::InvalidParam)
             }
         }
 
+        pub fn update_configured_counter(
+            counter_index: u64,
+            config_flags: PmuCounterConfigFlags,
+            event_type: PmuEventType,
+            active_cpu: &mut VmCpu,
+        ) {
+            let pmu_info = PmuInfo::get();
+            if pmu_info.get_counter_info(counter_index).is_some() {
+                active_cpu.set_counter_configuration(counter_index, event_type.raw());
+                if config_flags.get_auto_start() {
+                    active_cpu.set_started_counter(counter_index);
+                }
+            }
+        }
+
+        fn update_started_counters(
+            counter_index_base: u64,
+            counter_mask: u64,
+            active_cpu: &mut VmCpu,
+        ) {
+            let pmu_info = PmuInfo::get();
+            let num_counters = pmu_info.get_num_counters();
+            let configured_counters = active_cpu.get_configured_counters();
+            for i in counter_index_base..num_counters {
+                if configured_counters & (1 << i) != 0 && (i & counter_mask) != 0 {
+                    active_cpu.set_started_counter(i);
+                }
+            }
+        }
+
+        fn update_stopped_counters(
+            counter_index_base: u64,
+            counter_mask: u64,
+            active_cpu: &mut VmCpu,
+        ) {
+            let pmu_info = PmuInfo::get();
+            let num_counters = pmu_info.get_num_counters();
+            let started_counters = active_cpu.get_started_counters();
+            if counter_index_base < num_counters && started_counters != 0 {
+                for i in
+                    counter_index_base.max(started_counters.trailing_zeros().into())..num_counters
+                {
+                    if started_counters & (1 << i) != 0 && (i & counter_mask) != 0 {
+                        active_cpu.set_stopped_counter(i);
```

Comment:
same comment as above. The sanity check should happen before the SBI stop call.

Reply:
You should clear the started counter map here as well. No ?

Reply:
The flag gets cleared in vm_cpu.rs:134. Basically, started_counters is a read-only copy of the state maintained by the VmCpu:

`/// Sets the internal state for started PMU counters.
    /// This is intended to be called after a successful SBI invocation for stopping PMU counters.
    pub fn set_stopped_counter(&mut self, counter_index: u64, _read_counter_value: bool) {
        if (counter_index as usize) < self.configured_counter_map.len() {
            self.started_counters &= !(1 << counter_index);
        }
    }
`

Reply:
Ahh I missed that. Thanks.